src/nvim/popupmnu.c

```diff
@@ -95,19 +98,25 @@ void pum_display(pumitem_T *array, int size, int selected, bool array_changed)
     above_row = 0;
     below_row = cmdline_row;
 
-    // anchor position: the start of the completed word
-    row = curwin->w_wrow;
-    if (curwin->w_p_rl) {
-      col = curwin->w_width - curwin->w_wcol - 1;
+    if (State == CMDLINE) {
+      row = ui_has(kUICmdline) ? 0 : cmdline_row;
+      col = startcol;
+      pum_anchor_grid = ui_has(kUICmdline) ? -1 : DEFAULT_GRID_HANDLE;
```

Comment:
I would maybe parameterize `row`, `col`, and `pum_anchor_grid`, and then add a wrapper function `cmdline_pum_display()` in `ex_getln.c`, to avoid some of these decisions in `pum_display()`. Then I guess another wrapper is needed, `insert_pum_display()`. Maybe it's hopeless :)

Reply:
These wrapper functions already exist, they are called `ins_compl_show_pum` and `ex_getln.c:showmatches`. So we can move the code there and remove the `if/else` here.

Reply:
@justinmk Unfortunately this cannot work, `pum_set_selected` can change the window layout which causes `pum_display()` to be reexecuted (with an internal `do { } while(1)` ) so `ins_compl_show_pum` cannot precompute the positions.